doc/dev-guide/reconciler.md

```diff
@@ -0,0 +1,59 @@
+# Operator design
+
+## Controllers and their managed resources
+
+### lvmcluster-controller
+
+- On receiving a valid LVMCluster CR, lvmcluster-controller reconciles the
+  following resource units for setting up [Topolvm](topolvm-repo) CSI and all
+  the supporting resources to use storage local to the node via Logical Volume
+  Manager (lvm)
+- *csiDriver*: Reconciles topolvm CSI Driver
+- *topolvmController*: Reconciles topolvm controller plugin
+- *lvmVG*: Reconciles volume groups from LVMCluster CR
+- *openshiftSccs*: Manages SCCs when the operator is run in Openshift
+  environment
+- *topolvmNode*: Reconciles topolvm nodeplugin along with lvmd
+- *vgManager*: Responsible for creation of Volume Groups
+- *topolvmStorageClass*: Manages storage class life cycle based on
+  devicesClasses in LVMCluster CR
+- The LVMO creates an LVMVolumeGroup CR for each deviceClass in the
```

Comment:
This is the point I still don't understand... Why doesn't the admin create the LVMVolumeGroup CR directly? I don't see any description of what we're solving by embedding them in the cluster CR.

Reply:
- same as https://github.com/red-hat-storage/lvm-operator/pull/100#discussion_r798198723
- it's mostly about having to update status on the CR by multiple controllers which'll result into parallel CR update requests

Reply:
We want the operator to have the ability to generate the LVMVolumeGroup CRs based on the LVMCluster CR. This allows us better control and validation as discussed earlier.

Reply:
> it's mostly about having to update status on the CR by multiple controllers which'll result into parallel CR update requests

The problem of updating the status seems the same either way. I would agree that the operator should own updating the status on the LVMVolumeGroup CRs, but that doesn't mean the admin can't create those CRs.

> We want the operator to have the ability to generate the LVMVolumeGroup CRs based on the LVMCluster CR. This allows us better control and validation as discussed earlier.

What control and validation is needed? These are the details I'm missing. It seems any meaningful validation must happen on nodes by the vgmanager, not in the operator. The operator doesn't have insight to storage actually available.

Reply:
It could be something as simple as making sure there are not 2 entries trying to use up all the disks on all the nodes. It allows the operator to perform a first level validation before the vgmanager controller starts to create the VGs on the nodes. There could be other validations that we may perform in the future as the operator grows.

Reply:
There are multiple nodes and if multiple CRs are created at the same time, can we guarantee that all the nodes will process them in the same order?
If different nodes process the CRs in a different order, we cannot know which VGs can exist on which nodes. As the Storageclasses will have been created by then, potentially PVs can be created and then cleaning up becomes difficult. Hence I want to have a level of control on this.

Reply:
The API server will notify controllers about the CRs in the order of creation, so I expect the ordering will be consistent. 

As discussed on the call, I'm ok to keep this behavior for now until we give this more time to bake. In 4.10 it's anyway simplified with a single VG supported. I also don't see upgrades being affected negatively if we do decide later to remove the VGs from the LVMCluster CR. The LVMVolumeGroup CRs would have already been created by the operator, and the operator could just ignore the VGs in the LVMCluster after that.

I can see if that if we have compelling validation error-checking scenarios for clusters with multiple VGs we could keep this design. But fundamentally I still recommend not including the VGs in the LVMCluster CR for these reasons:
1. Bundling multiple resources into a single CR and forcing it to be a singleton is not a common K8s pattern I would expect on CRDs. A volume group is a logical unit, and therefore each VG makes sense as a single CR instance.  
2. Validation must be done on the nodes anyway by the vgmanager where the devices can be fully validated. The operator just can't fully validate available devices.
3. Instead of assuming multiple VGs would cause conflicts as an error condition, allow the first VG to consume the device. The later VG will simply not find the available device. The later VG won't be in error, it just might not have devices available to fulfill storage requests.

Reply:
As discussed on the call, I do not see any advantage to having the admin create the LVMVolumeGroup CRs. The actual interface exposed to the admin is the deviceClass - that is what we are internally mapping to lvm volume groups. We could decide to change this completely in the future.
The additional level of validation is something that makes things easier with setup especially when targeting hundreds of nodes. The node level validation done by the Vg manager is a different level of validation.

``` A volume group is a logical unit, and therefore each VG makes sense as a single CR instance.```
Yes, but that is not something the user needs to know. As far as they are concerned they get dynamic local storage and the exposing the lower layers to them is not really required. The LVMCluster CR will be fine tuned in future to provide this experience.

```Instead of assuming multiple VGs would cause conflicts as an error condition, allow the first VG to consume the device. The later VG will simply not find the available device. The later VG won't be in error, it just might not have devices available to fulfill storage requests.```

If there are failures in reconciliation for one VG or the reconcile requests are, for some reason, sent in a different order (network issues etc), this can lead to different VGs on different nodes. Since cleaning up is not easy, it makes more sense to prevent such conflicts as early as possible.

As discussed, this is the approach we are following now. Any further decisions will be based on user experience and feedback and will be taken up later.